### PR TITLE
Add name and address filter in Android

### PR DESCRIPTION
**Pull request by PavlosTze**
Implements #377

### CODE REVIEW

File: core/src/androidMain/kotlin/Filter.kt
Comment (twyatt) [inline core/src/androidMain/kotlin/Filter.kt:31]: I think it makes sense to have these as individual classes so that their values don't need to be `null`able.

```suggestion
    /**
     * https://developer.android.com/reference/android/bluetooth/le/ScanFilter.Builder#setDeviceName(java.lang.String)
     */
    public actual data class Name actual constructor(
        public val name: String,
    ) : Filter()

    /**
     * https://developer.android.com/reference/android/bluetooth/le/ScanFilter.Builder#setDeviceAddress(java.lang.String)
     */
    public data class Address(
        public val address: String,
    ) : Filter()
```

**Review by twyatt — COMMENTED** [review]
It'd be great to implement the filtering (for both "name" and "address") for all platforms (JS and Apple as well) — for Apple, the filtering would have the be "emulated" (since Core Bluetooth doesn't provide filtering of name nor address via their provided APIs).

@PavlosTze do you want to take that on, or would you like me to add some commits to your branch, or create a new PR on top of this one?

**Comment by PavlosTze** [comment]
I made the fixes that were needed to use 2 seperate classes for `Name` and `Address`. Would it be possible @twyatt to add some commits either on my branch, or create a new PR on top of this one for the JS & Apple functionality?

I'm having some trouble to understand how the JS filtering works (as I see the option for having a name filter already there) and I'm not familiar with Apple development at all.

**Comment by twyatt** [comment]
> Would it be possible @twyatt to add some commits either on my branch, or create a new PR on top of this one for the JS & Apple functionality?
> 
> I'm having some trouble to understand how the JS filtering works (as I see the option for having a name filter already there) and I'm not familiar with Apple development at all.

No worries. I'll try to find some time soon to implement the JS and Apple side of this. 👍

**Comment by PavlosTze** [comment]
@twyatt Any news to that?

**Comment by twyatt** [comment]
Apologies, life has been a bit chaotic but I am hoping things will calm in the coming weeks, and I'll be able to put focus back on Kable. Thanks so much for being patient.

**Comment by twyatt** [comment]
I've begun working on this on a local branch:

- [x] Migrate filters to common code
- [x] Support scan filtering on Apple
    - [ ] Test
- [ ] Support scan filter on JS
    - [ ] Test

**Comment by twyatt** [comment]
My local branch ended up being a major rework/rewrite of the filtering for all platforms. It was based on (continuation of) your work in this PR, so my plan now is the get this PR merged and the branch I'm currently working on will becoming another follow up PR.

**Comment by PavlosTze** [comment]
That sounds great @twyatt !

**Review by twyatt — APPROVED** [review]

**Comment by twyatt** [comment]
2 approvals are needed for merge, but our offices are closed until Thursday; but as soon as I get another approval I'll get this merged.

I'll also link to the follow up PR once I've published it (so you'll get to see how the filtering is changing — which from a public API perspective shouldn't be too drastic).

**Comment by twyatt** [comment]
@PavlosTze https://github.com/JuulLabs/kable/pull/408 is in draft but should give you an idea of how the filtering implementation will be changing.

**Review by davidtaylor-juul — APPROVED** [review]

**Pull request merged**
